packages/docusaurus-theme-classic/src/theme/Footer/index.tsx

```diff
@@ -6,185 +6,27 @@
-                    __html: item.html,
-                  }}
-                />
-              ) : (
-                <li key={item.href || item.to} className="footer__item">
-                  <FooterLink {...item} />
-                </li>
-              ),
-            )}
-          </ul>
-        </div>
-      ))}
-    </>
-  );
-}
-
-function SimpleLinks({links}: {links: SimpleFooter['links']}) {
-  return (
-    <div className="footer__links">
-      {links.map((item, key) => (
-        <React.Fragment key={key}>
-          {item.html ? (
-            <span
-              className="footer__link-item"
-              // Developer provided the HTML, so assume it's safe.
-              // eslint-disable-next-line react/no-danger
-              dangerouslySetInnerHTML={{
-                __html: item.html,
-              }}
-            />
-          ) : (
-            <FooterLink {...item} />
-          )}
-          {links.length !== key + 1 && (
-            <span className="footer__link-separator">·</span>
-          )}
-        </React.Fragment>
-      ))}
-    </div>
-  );
-}
-
-function isMultiColumnFooterLinks(
-  links: MultiColumnFooter['links'] | SimpleFooter['links'],
-): links is MultiColumnFooter['links'] {
-  return 'title' in links[0]!;
-}
+import {useThemeConfig} from '@docusaurus/theme-common';
+import FooterLinks from '@theme/Footer/Links';
+import FooterLogo from '@theme/Footer/Logo';
+import FooterCopyright from '@theme/Footer/Copyright';
+import FooterLayout from '@theme/Footer/Layout';
 
 function Footer(): JSX.Element | null {
   const {footer} = useThemeConfig();
   if (!footer) {
     return null;
   }
-  const {copyright, links, logo} = footer;
+  const {copyright, links, logo, style} = footer;
 
   return (
-    <footer
-      className={clsx('footer', {
-        'footer--dark': footer.style === 'dark',
-      })}>
-      <div className="container container-fluid">
-        {links &&
-          links.length > 0 &&
-          (isMultiColumnFooterLinks(links) ? (
-            <div className="row footer__links">
-              <MultiColumnLinks links={links} />
-            </div>
-          ) : (
-            <div className="footer__links text--center">
-              <SimpleLinks links={links} />
-            </div>
-          ))}
-        {(logo || copyright) && (
-          <div className="footer__bottom text--center">
-            <FooterLogo logo={logo} />
-            {copyright && (
-              <div
-                className="footer__copyright"
-                // Developer provided the HTML, so assume it's safe.
-                // eslint-disable-next-line react/no-danger
-                dangerouslySetInnerHTML={{
-                  __html: copyright,
-                }}
-              />
-            )}
-          </div>
-        )}
-      </div>
-    </footer>
+    <FooterLayout
+      style={style}
+      links={links && links.length > 0 && <FooterLinks links={links} />}
```

Comment:
I don't really like this style... Can we just pass `links` down and render it in `FooterLayout`?

I know that it means `FooterLayout` will now be tied to those sub-components. TBH, I'm not even clear what the value of this abstraction is, because users will likely not swizzle/change this generic component's props without altering the other ones' anyways.

Reply:
It is important for me to isolate the macro layout of a theme component in its own abstraction, mostly taking `ReactNode` props.

It's also a good example of what I'd like all theme "big sections" to look like.

Users should be able to easily do things like put the logo/copyright above the links, add extra sections etc without managing much code complexity (ie dealing with simple vs multi-column layouts and type guards). 

I agree it may look a bit overkill though (as the code complexity it hides is not so complex) but I'd like to keep it this way, as an example, and for consistency with what we'll do in other places (where it will be less overkill)

Reply:
Okay, I see 👍

Should we move the `links && links.length > 0 && ` to `FooterLinks` then? This will make this component cleaner and also more robust.

Reply:
Doing so is not the best idea and could lead to antipatterns. In general it's preferable if the component itself doesn't decide if it renders something or not, because it makes a composition with layouts more complex and error-prone.

For example, look at the logo here, we add a margin if the logo is defined:

<img width="457" alt="image" src="https://user-images.githubusercontent.com/749374/157861234-68d3a021-4228-4b0b-854d-cbb2bf34f33a.png">


Now if we always pass a ReactNode as props, the prop is always defined and this potentially leads to a margin being added uselessly around a null element.

In general, I'd like to avoid components rendering null (preferring the parent to not render the component in the first place) unless it's clear it's safe to do so. For example, it looks quite safe for the top-level Footer component to render null

Reply:
`links && links.length > 0 && <FooterLinks links={links} />` where `FooterLinks` evaluates to `null` is the same as `links` being `null`, no? 🤷‍♂️

Reply:
Not sure to understand what you mean here 😅 

Imagine the user swizzling the layout wants to add some margin around the links, he should be able to do so:

```jsx
{links && <ExtraMargin>{links}</ExtraMargin>}
```

Now if `links` is always a ReactNode in the layout component, then the extra margin is always added, even if links is null

Reply:
I see... Okay, makes sense

Reply:
👍 

going to merge then

if you have improvement ideas, don't hesitate to submit PRs before 2.0